---
title: "stat547-hw8-thibodeau-mylinh-process-reporting"
author: "My Linh Thibodeau"
date: '2017-11-17'
output: github_document
---

```{r}
suppressMessages(suppressWarnings(library(plyr)))
suppressMessages(suppressWarnings(library(tidyverse)))
suppressMessages(suppressWarnings(library(stringr)))
```

# PURPOSE OF THIS FILE: Dataframe manipulation and joining

# SOURCES OF DATA

References for data used in this homework.

* DECIPHER: Database of Chromosomal Imbalance and Phenotype in Humans using Ensembl Resources. Firth, H.V. et al (2009). Am.J.Hum.Genet 84, 524-533 (DOI: dx.doi.org/10/1016/j.ajhg.2009.03.010). 
* This homework makes use of data generated by the DECIPHER community. A full list of centres who contributed to the generation of the data is available from http://decipher.sanger.ac.uk and via email from decipher@sanger.ac.uk. Funding for the project was provided by the Wellcome Trust.
* Gray KA, Yates B, Seal RL, Wright MW, Bruford EA. genenames.org: the HGNC resources in 2015. Nucleic Acids Res. 2015 Jan;43(Database issue):D1079-85. doi: 10.1093/nar/gku1071. PMID:25361968. 
* HGNC Database, HUGO Gene Nomenclature Committee (HGNC), EMBL Outstation - Hinxton, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridgeshire, CB10 1SD, UK www.genenames.org (complete HGNC dataset downloaded [here](https://www.genenames.org/cgi-bin/statistics) on November 17/2017).
* Orphanet: an online database of rare diseases and orphan drugs. Copyright, INSERM 1997. Available at http://www.orpha.net Accessed (accessed November 17, 2017).
* Orphadata: Free access data from Orphanet. Â© INSERM 1997. Available on http://www.orphadata.org. Data version (XML data version en_product4_HPO.xml).


## (1) Getting and formating the data

#### Load the data

```{r}
decipher_DDG2P <- read.table("raw_data/DDG2P_17_11_2017.csv", header = TRUE, sep =",")
hgnc_phenotype <- read.table("raw_data/hgnc_phenotype.tsv", header = TRUE, sep="\t")
hgnc_data <- read.table("raw_data/hgnc_data.tsv", sep="\t", header=TRUE, fill=TRUE)
orphadata_phenotype <- read.table("raw_data/orphadata_phenotypes_MLT.csv", sep=",", header = TRUE, strip.white = TRUE, row.names=NULL, fill=TRUE, colClasses = c("phenotype_id"= "character", "disorder_name" = "character", "phenotype_name" = "character", "frequency" = "character")) %>% select(phenotype_id, disorder_name, phenotype_name, frequency)

# Remove empty values
orphadata_phenotype <- orphadata_phenotype %>% filter(phenotype_id !="")
orphadata_phenotype <- orphadata_phenotype %>% filter(disorder_name !="")
orphadata_phenotype <- orphadata_phenotype %>% filter(phenotype_name != "")
orphadata_phenotype <- orphadata_phenotype %>% filter(frequency != "")

# load and clean orphadata genes
orphadata_genes <- read.table("raw_data/orphadata_genes.tsv", sep="\t", header=TRUE, fill= TRUE)
orphadata_genes <- orphadata_genes %>% filter(Disorder.OrphaNumber != "")
orphadata_genes <- orphadata_genes %>% filter(Gene.symbol != "")

# View commands if needed
# View(decipher_DDG2P)
# View(hgnc_phenotype)
# View(hgnc_data)
# View(orphadata_phenotype)
# View(orphadata_genes)
```

#### Change disease/disorder names to lower cases 

```{r}
decipher_DDG2P <- decipher_DDG2P %>%
	mutate(disease.name = str_to_lower(disease.name))
# View(decipher_DDG2P)
orphadata_phenotype <- orphadata_phenotype %>%
	mutate(disorder_name = str_to_lower(disorder_name))
# View(orphadata_phenotype)
```

#### Writing a short version of orphadata_genes to a file

```{r}
orphadata_genes_subset <- orphadata_genes %>%
	select(Disorder.name, Disorder.OrphaNumber, Gene.symbol)
```

```{r}
write_tsv(orphadata_genes_subset, "~Desktop/STAT545-HW-thibodeau-mylinh/hw08-shiny-appdata/orphadata_genes_subset.tsv")
```


#### Joining orphadata_phenotype and orphadata_genes_subset

Also make orphadata_genes_subset disorder name lower case.
```{r}
orphadata_genes_subset <- orphadata_genes_subset %>%
	mutate(Disorder.name = str_to_lower(Disorder.name))
# View(orphadata_genes_subset)
orphadata_genes_phenotype_subset <- orphadata_phenotype %>% 
	left_join(orphadata_genes_subset, by = c("disorder_name"="Disorder.name"))
# View(orphadata_genes_phenotype_subset)
```
Note. This dataset will not be used for now.

#### Joining datasets

```{r}
decipher_DDG2P_hgnc_phenotype <- decipher_DDG2P %>% 
	left_join(hgnc_phenotype, by = c("gene.symbol"="symbol"))
# View(decipher_DDG2P_hgnc_phenotype)
# colnames(decipher_DDG2P_hgnc_phenotype) %>% View()

decipher_DDG2P_hgnc_phenotype <- decipher_DDG2P_hgnc_phenotype %>%
	select(gene.symbol, gene.mim, disease.name, disease.mim, allelic.requirement, DDD.category, mutation.consequence, phenotypes, organ.specificity.list, mutation.consequence, hgnc.id)
# View(decipher_DDG2P_hgnc_phenotype)

decipher_DDG2P_hgnc_phenotype_plus <- decipher_DDG2P_hgnc_phenotype %>% left_join(hgnc_data, by = c("gene.symbol"="symbol"))
# View(decipher_DDG2P_hgnc_phenotype_plus)
# colnames(decipher_DDG2P_hgnc_phenotype_plus) %>% View()

# Make disease.name all lower case
decipher_DDG2P_hgnc_phenotype_plus <- decipher_DDG2P_hgnc_phenotype_plus %>%
	mutate(disease.name = str_to_lower(disease.name))

decipher_DDG2P_hgnc_phenotype_subset <- decipher_DDG2P_hgnc_phenotype_plus[1:22]
decipher_DDG2P_hgnc_phenotype_subset <- decipher_DDG2P_hgnc_phenotype_subset %>%
	select(-phenotypes, -hgnc_id)
# View(decipher_DDG2P_hgnc_phenotype_subset) 

DDG2P_hgnc_orphadata <- orphadata_phenotype %>%
	right_join(decipher_DDG2P_hgnc_phenotype_subset, by = c("disorder_name" = "disease.name")) 
# View(DDG2P_hgnc_orphadata)

# DDG2P_hgnc_orphadata <- DDG2P_hgnc_orphadata
DDG2P_hgnc_orphadata <- DDG2P_hgnc_orphadata %>% as.data.frame()

# dim(DDG2P_hgnc_orphadata)
# summary(DDG2P_hgnc_orphadata) %>% View()
# View(DDG2P_hgnc_orphadata)
```

Writing DDG2P_hgnc_orphadata to a file for the app to use

```{r}
write_tsv(DDG2P_hgnc_orphadata, "~Desktop/STAT545-HW-thibodeau-mylinh/hw08-shiny-appdata/DDG2P_hgnc_orphadata.tsv")
```
Note. I tried to add the split organs data, but the file was massive and exceeded the size limit allowed.


## Additional helpful "short" lists

### List of possible organs involved

**We need to start by spliting the organ.specificity.list into individual components.**

First, to preserve the association between the gene and the organs, I need to create a different dataframe with the genes and the organ.specificity.list. 

```{r}
genes_organs <- DDG2P_hgnc_orphadata %>%
	ungroup() %>%
 	select(gene.symbol, organ.specificity.list) %>%
	mutate(gene_symbol_organ = str_c(gene.symbol, organ.specificity.list, sep=";"))
# View(genes_organs)
genes_organs_split <- 
	str_split_fixed(genes_organs$gene_symbol_organ, ";" , n = 9) %>% as.data.frame() %>%
	unique()
# View(genes_organs_split)
```


Second, let's store the list of possible organs affected.
```{r}
organs_choices <- genes_organs_split %>%
	select(-V1) %>%
	gather() %>%
	select(value) %>%
	unique()
colnames(organs_choices) <- c("organ_types")
organs_choices <- organs_choices %>% arrange(organ_types)
```

Write a "short" list of possible organs involved to a file.
```{r}
write_tsv(organs_choices, "~Desktop/STAT545-HW-thibodeau-mylinh/hw08-shiny-appdata/organs_choices.tsv")
```

Third, let's attribute column names to genes_organs_split and 
```{r}
colnames(genes_organs_split) <- (c("gene.symbol", "organ1", "organ2", "organ3", "organ4", "organ5", "organ6", "organ7", "organ8")) 
write_tsv(genes_organs_split, "~Desktop/STAT545-HW-thibodeau-mylinh/hw08-shiny-appdata/genes_organs_split.tsv")
```

Let's make a "gathered" version of this dataframe with one column containing the gene name and the second column containing all the potential organs affected for that gene.
```{r}
genes_organs_gather <- genes_organs_split %>%
	gather(key = organ_number, value = organs_affected, organ1:organ8)
genes_organs_gather <- genes_organs_gather %>%
	select(gene.symbol, organs_affected) %>%
	filter(organs_affected != "")
# genes_organs_gather <- genes_organs_gather %>% 
	# group_by(gene.symbol)
# View(genes_organs_gather)
```

Finally, let's write this new formated data to a file
```{r}
write_tsv(genes_organs_gather, "~Desktop/STAT545-HW-thibodeau-mylinh/hw08-shiny-appdata/genes_organs_gather.tsv")
```

### I will also used this genes_organs_gather to merge with the DDG2P_hgnc_orphadata dataframe


```{r}
DDG2P_hgnc_orphadata <- DDG2P_hgnc_orphadata %>% ungroup()
DDG2P_hgnc_orphadata_organs_listed <-
		genes_organs_gather %>%
		ungroup() %>%
		full_join(DDG2P_hgnc_orphadata, by = c("gene.symbol" = "gene.symbol"))
DDG2P_hgnc_orphadata_organs_listed <- as.data.frame(DDG2P_hgnc_orphadata_organs_listed)
DDG2P_hgnc_orphadata_organs_listed <- DDG2P_hgnc_orphadata_organs_listed[1:14]
# dim(DDG2P_hgnc_orphadata_organs_listed)
# View(DDG2P_hgnc_orphadata_organs_listed)
```
Note. I only kept the 14 first columns as they are most informative. The file is massive, and I wonder if that is why my app has been running slowly. I initially did this data joining inside the app, but I realize now that it was not a good idea.

And write it to a file
```{r}
write_tsv(DDG2P_hgnc_orphadata_organs_listed, "~Desktop/STAT545-HW-thibodeau-mylinh/hw08-shiny-appdata/DDG2P_hgnc_orphadata_organs_listed.tsv")
```

### List of possible phenotype_name

List the unique phenotype_name entries
```{r}
orphadata_phenotype_choices <- orphadata_phenotype %>% select(phenotype_name) %>% unique() 
```

Write the result to a file
```{r}
write_tsv(orphadata_phenotype_choices, "~Desktop/STAT545-HW-thibodeau-mylinh/hw08-shiny-appdata/orphadata_phenotype_choices.tsv")
```


