# README -  stat547-hw8-thibodeau-mylinh

In progress

## SOURCE OF DATA

References for data used in this homework.

* DECIPHER: Database of Chromosomal Imbalance and Phenotype in Humans using Ensembl Resources. Firth, H.V. et al (2009). Am.J.Hum.Genet 84, 524-533 (DOI: dx.doi.org/10/1016/j.ajhg.2009.03.010). 
* This homework makes use of data generated by the DECIPHER community. A full list of centres who contributed to the generation of the data is available from http://decipher.sanger.ac.uk and via email from decipher@sanger.ac.uk. Funding for the project was provided by the Wellcome Trust.
* Gray KA, Yates B, Seal RL, Wright MW, Bruford EA. genenames.org: the HGNC resources in 2015. Nucleic Acids Res. 2015 Jan;43(Database issue):D1079-85. doi: 10.1093/nar/gku1071. PMID:25361968. 
* HGNC Database, HUGO Gene Nomenclature Committee (HGNC), EMBL Outstation - Hinxton, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridgeshire, CB10 1SD, UK www.genenames.org (complete HGNC dataset downloaded [here](https://www.genenames.org/cgi-bin/statistics) on November 17/2017).
* Orphanet: an online database of rare diseases and orphan drugs. Copyright, INSERM 1997. Available at http://www.orpha.net Accessed (accessed November 17, 2017).
* Orphadata: Free access data from Orphanet. Â© INSERM 1997. Available on http://www.orphadata.org. Data version (XML data version en_product4_HPO.xml).


# FILES TO REVIEW

I experienced many challenges (see section below) during this homework. 

* My Shiny App is stored locally and not on Github, but I copied the files into [this folder]()
	- File [User Interface (UI)]()
	- File [Server]()
	- File [data folder]()
* Since the topic is specialized (medical genetics/genomics), I made a [tutorial]() for you to text my app.

Extensive data manipulation/transformation was required before even obtaining the dataset required for my app, so I have a dedicated Rmd/Md file for data manipulation [here]().

## Some issues encoutered during this homework

#### (1) Data manipulation with csv and tab separated files and `read.table()` function

I discovered that `read.table()` does NOT like the presence of apostrophes (') in the data at all. It didn't parse the columns appropriately. The `read.csv()` and `read_csv()` did not work out on the messy datasets. 

I have to remove the apostrophes in a non-programming-kosher way (basically in TextWrangler) in order to be able to get going. Unfortunately, it took me a lot of time before being able to figure out the problem, so I wasted a lot of time in trying to do data manipulation. 

#### (2) XML file format - Too tricky for me

Luckily, a friend of mine helped me to get the data from orphanet/orphadata xml files. I tried to manipulate the files myself with the XML and xml2 packages, but was unable to reach completion.

#### (3) Ways to make messy data even messier

Yes, that is true, I experienced this phenomenon in this homework. Here are some "simple" ways to mess your data even more (so basically things you should pay attention to):

1. Matrix vs dataframe: In data table manipulation, I occasionally found myself handling matrices rather than dataframe and I didn't even know it happened.
2. Blank vs NA: using different file formats (csv and tsv, and also at some point txt), I realized that some "empty cells" of my dataframes were being populated by "NA" and some others remained blank. I will let you imagine the numerous problems that can arise with trying to filter that kind of datasets based on a feature.
3. Small glitches made big: If there is a small glitch in one dataset, and you join it to other datasets and perform some data manipulation steps to obtain a single final dataframes, the glitches become magnified by the effect of manipulating the data. 

***
## Newly acquired skills or knowledge

Well, basically everything I have done for this Shiny App homework was new, but I will still list a few things.

1. Always verify the class of your data (e.g. list, dataframe, matrix).
2. The Shiny app tutorial [here](https://shiny.rstudio.com/tutorial/) is really good, I highly recommend it !
3. I learned how to make a shiny app ! You can see my app directly to know the technical aspects, but here are some examples:
	- tags$ are a family of functions (e.g. `tags$h1()` for header 1, `tags$em()` for italic writing, etc.) that allows to modify text output. The most frequently used have been made into functions (e.g. `tags$h1()` is the same thant `h1()`)
	- `tabsetPanel()` with `tabPanel()` subdivisions allows to create multiple tabs for the user to navigate
	- `actionButton()` to make a clickable button, and also how to use eventReactive() so that the table output will not be displayed until the use clicks on the button.
	- I have learned how to use these functions as well: radioButtons(), wellPanel(), fluidRow(column())
4. Here are some things I learned that my app won't tell you:
	- The flow between the UI and Server files
	- That it's a good idea to annotate your code extensively, because when reading it back, if it doesn't make sense to me, it will be impossible to read for somebody else :no_mouth:
	- Indentation is so important !! But sometimes, I found that the default indentation is visually confusing and difficult to read (especially when trying to find a missing parenthesis), but it's better to stick to the default so that it's consistent.

Thank you for your time,
My Linh

***
### Other references

* Code chunks in part from the Shiny RStudio website [here](http://shiny.rstudio.com/articles/app-formats.html)
* R and XML package "how to use" [here](https://www.r-bloggers.com/r-and-the-web-for-beginners-part-ii-xml-in-r/) and [here](https://stackoverflow.com/questions/19747488/parsing-xml-in-r)