---
title: "stat547-hw8-thibodeau-mylinh-process-reporting"
author: "My Linh Thibodeau"
date: '2017-11-17'
output: github_document
---

```{r}
suppressMessages(suppressWarnings(library(plyr)))
suppressMessages(suppressWarnings(library(tidyverse)))
suppressMessages(suppressWarnings(library(stringr)))
```

# SOURCES OF DATA

References for data used in this homework.

* DECIPHER: Database of Chromosomal Imbalance and Phenotype in Humans using Ensembl Resources. Firth, H.V. et al (2009). Am.J.Hum.Genet 84, 524-533 (DOI: dx.doi.org/10/1016/j.ajhg.2009.03.010). 
* This homework makes use of data generated by the DECIPHER community. A full list of centres who contributed to the generation of the data is available from http://decipher.sanger.ac.uk and via email from decipher@sanger.ac.uk. Funding for the project was provided by the Wellcome Trust.
* Gray KA, Yates B, Seal RL, Wright MW, Bruford EA. genenames.org: the HGNC resources in 2015. Nucleic Acids Res. 2015 Jan;43(Database issue):D1079-85. doi: 10.1093/nar/gku1071. PMID:25361968. 
* HGNC Database, HUGO Gene Nomenclature Committee (HGNC), EMBL Outstation - Hinxton, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridgeshire, CB10 1SD, UK www.genenames.org (complete HGNC dataset downloaded [here](https://www.genenames.org/cgi-bin/statistics) on November 17/2017).
* Orphanet: an online database of rare diseases and orphan drugs. Copyright, INSERM 1997. Available at http://www.orpha.net Accessed (accessed November 17, 2017).
* Orphadata: Free access data from Orphanet. Â© INSERM 1997. Available on http://www.orphadata.org. Data version (XML data version en_product4_HPO.xml).


## (1) Getting and formating the data

#### Load the data

```{r}
decipher_DDG2P <- read.table("raw_data/DDG2P_17_11_2017.csv", header = TRUE, sep =",")
hgnc_phenotype <- read.table("raw_data/hgnc_phenotype.tsv", header = TRUE, sep="\t")
hgnc_data <- read.table("raw_data/hgnc_data.tsv", sep="\t", header=TRUE, fill=TRUE)
orphadata_phenotype <- read.table("raw_data/orphadata_phenotypes.csv", sep=",", header = TRUE, fill=TRUE) %>% select(phenotype_id, disorder_name, phenotype_name, diagnostic_criteria, frequency)
# View(decipher_DDG2P)
# View(hgnc_phenotype)
# View(hgnc_data)
# View(orphadata_phenotype)
```

#### Change disease/disorder names to lower cases 

```{r}
decipher_DDG2P <- decipher_DDG2P %>%
	mutate(disease.name = str_to_lower(disease.name))
orphadata_phenotype <- orphadata_phenotype %>%
	mutate(disorder_name = str_to_lower(disorder_name))
```


#### Joining datasets

```{r}
decipher_DDG2P_hgnc_phenotype <- decipher_DDG2P %>% 
	left_join(hgnc_phenotype, by = c("gene.symbol"="symbol"))
# View(decipher_DDG2P_hgnc_phenotype)
# colnames(decipher_DDG2P_hgnc_phenotype)
decipher_DDG2P_hgnc_phenotype <- decipher_DDG2P_hgnc_phenotype %>%
	select(gene.symbol, gene.mim, disease.name, disease.mim, allelic.requirement, DDD.category, mutation.consequence, phenotypes, organ.specificity.list, mutation.consequence, hgnc.id)
decipher_DDG2P_hgnc_phenotype_plus <- decipher_DDG2P_hgnc_phenotype %>% left_join(hgnc_data, by = c("gene.symbol"="symbol"))
# View(decipher_DDG2P_hgnc_phenotype_plus)
# colnames(decipher_DDG2P_hgnc_phenotype_plus)
decipher_DDG2P_hgnc_phenotype_subset <- decipher_DDG2P_hgnc_phenotype_plus[1:22]
# View(decipher_DDG2P_hgnc_phenotype_subset) 
DDG2P_hgnc_orphadata <- decipher_DDG2P_hgnc_phenotype_subset %>%
	select(-hgnc_id, -prev_name) %>%
	left_join(orphadata_phenotype, by = c("disease.name" ="disorder_name")) %>% group_by(disease.name)
# View(DDG2P_hgnc_orphadata)
```


#### Split the organ.specificity.list into individual components

First, to preserve the association between the gene and the organs, I need to create a different dataframe with the genes and the organ.specificity.list. 

```{r}
genes_organs <- DDG2P_hgnc_orphadata %>%
	ungroup() %>%
 	select(gene.symbol, organ.specificity.list) %>%
	mutate(gene_symbol_organ = str_c(gene.symbol, organ.specificity.list, sep=";"))
# View(genes_organs)
genes_organs_split <- 
	str_split_fixed(genes_organs$gene_symbol_organ, ";" , n = 9) %>% as.data.frame()
# View(genes_organs_split)
```

Second, let's store the list of possible organs affected.
```{r}
organs_choices <- genes_organs_split %>%
	select(-V1) %>%
	gather() %>%
	select(value) %>%
	unique()
colnames(organs_choices) <- c("organ_types")
write_tsv(organs_choices, "~/Desktop/STAT545-HW-thibodeau-mylinh/stat547-hw8-thibodeau-mylinh/genomic_disorders_quiz/data/organs_choices.tsv")
```

Third, let's attribute column names to genes_organs_split and then join this data to the DDG2P_hgnc_orphadata dataframe.
```{r}
colnames(genes_organs_split) <- (c("gene.symbol", "organ1", "organ2", "organ3", "organ4", "organ5", "organ6", "organ7", "organ8"))
DDG2P_hgnc_orphadata_v2 <- DDG2P_hgnc_orphadata %>%
	left_join(genes_organs_split, by = c("gene.symbol" = "gene.symbol"))
# View(DDG2P_hgnc_orphadata_v2)
```


## (2) Writing to a data file for the app to use

```{r}
write_tsv(DDG2P_hgnc_orphadata_v2, "genomic_disorders_quiz/data/DDG2P_hgnc_orphadata_v2.tsv")
```

