---
title: "stat545-hw04-thibodeau-mylinh"
author: "My Linh Thibodeau"
date: '2017-10-07'
output: github_document
---

```{r}
suppressPackageStartupMessages(library(tidyverse))
knitr::opts_chunk$set(fig.width=12, fig.height=9)
library(knitr)
library(kableExtra)
options(knitr.table.format = "html")
#install.packages("scales")
library(scales)
#install.packages("tidygenomics")
library(tidygenomics)
```

# Genomic dataset - A few clarifications

Vincenzo Coia has approved my request to use published genomic data for the next assignments, therefore, I want to provide a few clarifications:

* I have tried to introduce some basic explanations about the genomic dataset, but obviously, this is not a genetics course and my objective is to explore and learn how to use R and its packages, not to teach complex notions of cancer genomic analysis. Therefore, I don't expect people to understand what the data and plots represent if they haven't studied in related fields.  
* I would recommend you make abstraction of the underlying biological context and simply try to read this homework based on the variable types (e.g. copy.category is a categorical variable like "gain" or "loss") rather than what they represent.

## SOURCE OF DATA - supplementary files of published articles:

Thibodeau, M. L. et al. Genomic profiling of pelvic genital type leiomyosarcoma in a woman with a germline CHEK2:c.1100delC mutation and a concomitant diagnosis of metastatic invasive ductal breast carcinoma. Cold Spring Harb Mol Case Stud mcs.a001628 (2017). doi:10.1101/mcs.a001628
Open access article and data available [here](http://molecularcasestudies.cshlp.org/content/3/5/a001628.long)  

Jones, M. R. et al. Response to angiotensin blockade with irbesartan in a patient with metastatic colorectal cancer. Ann Oncol 27, 801–806 (2016). doi:10.1093/annonc/mdw060
Open access article and data available [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4843189/)

# Setting up the data

## Getting the data ready

Getting the CHEK2 paper Thibodeau et al. 2017 rna-seq and copy number variant (cnv) data and cleaning it up.

If you would like the details about the code chunk below, please refer to my homework3.

```{r}
chek2_rna_cnv <- read.table("/Users/mylinh/Documents/published_pog_data/stat545-tables/chek2-rna-cnv.txt", sep="\t",  strip.white = TRUE, header = TRUE)
chek2_rna_cnv <- with(chek2_rna_cnv, chek2_rna_cnv[!(chr==""),])
chek2_rna_cnv <- with(chek2_rna_cnv, chek2_rna_cnv[!(start==""),])
chek2_rna_cnv <- with(chek2_rna_cnv, chek2_rna_cnv[!(end==""),])
chek2_rna_cnv <- with(chek2_rna_cnv, chek2_rna_cnv[!(Ensembl.gene.ID==""),])
chek2_rna_cnv <- with(chek2_rna_cnv, chek2_rna_cnv[!(hugo==""),])
chek2_rna_cnv$cancer.gene.type <- as.character(chek2_rna_cnv$cancer.gene.type)
chek2_rna_cnv$cancer.gene.type[chek2_rna_cnv$cancer.gene.type==""] <- "NA"
chek2_rna_cnv$cancer.gene.type <- as.factor(chek2_rna_cnv$cancer.gene.type)
chek2_rna_cnv <- chek2_rna_cnv %>%
  group_by(hugo) %>%
  slice(1L)
gene_types <- levels(chek2_rna_cnv$cancer.gene.type)
abbreviations_gene_types  <- factor(c("NA", "ONC", "ONC;p-TS", "ONC;TS", "p-ONC", "p-ONC;p-TS", "p-ONC;TS", "p-TS", "TS"))
abbreviations_gene_types <- as.character(abbreviations_gene_types)
abbr_table <- data.frame(Abbreviation = abbreviations_gene_types, cancer.gene.type = gene_types)
abbr_table %>% kable("html") %>% kable_styling()
chek2_rna_cnv$cancer.gene.type <- as.character(chek2_rna_cnv$cancer.gene.type)
chek2_rna_cnv$cancer.gene.type <- gsub("oncogene", "ONC", chek2_rna_cnv$cancer.gene.type)
chek2_rna_cnv$cancer.gene.type <- gsub("tumour suppressor", "TS", chek2_rna_cnv$cancer.gene.type)
chek2_rna_cnv$cancer.gene.type <- gsub("putative", "p-", chek2_rna_cnv$cancer.gene.type)
chek2_rna_cnv$cancer.gene.type <- as.vector(chek2_rna_cnv$cancer.gene.type)
View(chek2_rna_cnv)
```

## Getting the other dataset ready

Getting the irbesartan-colorectal cancer paper Jones et al (2016) rna-seq and copy number variant (cnv) data and see what it looks like.

```{r}
arb_rna_cnv <- read.table("/Users/mylinh/Documents/published_pog_data/stat545-tables/crc-arb-response-rna-cnv-v2.txt", header = TRUE, sep="\t", fill = TRUE)
```
Note. Actually, I had to create a second file (v2) removing some columns, because when I was trying to read the first version of the file, RStudio kept crashing down. 

```{r}
summary(arb_rna_cnv)
class(arb_rna_cnv)
kable(sapply(arb_rna_cnv, class))
str(arb_rna_cnv)
dim(arb_rna_cnv)
```


```{r}
arb_rna_cnv <- with(arb_rna_cnv, arb_rna_cnv[!(chr==""),])
arb_rna_cnv <- with(arb_rna_cnv, arb_rna_cnv[!(pos==""),])
arb_rna_cnv <- with(arb_rna_cnv, arb_rna_cnv[!(ensid==""),])
arb_rna_cnv <- with(arb_rna_cnv, arb_rna_cnv[!(HUGO_ID==""),])
arb_rna_cnv <- arb_rna_cnv %>%
  group_by(hugo) %>%
  slice(1L)
dim(arb_rna_cnv)
View(arb_rna_cnv)
```

## TIDY GENOMICS

Tidyverse is an incredible package, and some scientists have expanded its concepts to genomics data. Constantine Ahlmann-Eltze's blog explained the use of the [Tidy Genomics package](https://cran.r-project.org/web/packages/tidygenomics/vignettes/intro.html), which I am hoping to have the opportunity of exploring during this homework, but we will see if my data is appropriate for that later. The peculiarity of genomics data is the need for accomodating ranges, and tidygenomics has its own functions for reshaping and aggregating data:

* genome_intersect
* genome_subtract
* genome_join_closest
* genome_cluster
* genome_complement
* genome_join

# General data reshaping and relationship to aggregation

Activity #1

Make you own cheatsheet similar to Tyler Rinker’s minimal guide to tidyr
