---
title: "stat545-hw04-thibodeau-mylinh"
author: "My Linh Thibodeau"
date: '2017-10-07'
output: github_document
---

```{r}
suppressPackageStartupMessages(library(tidyverse))
knitr::opts_chunk$set(fig.width=12, fig.height=9)
library(knitr)
library(kableExtra)
options(knitr.table.format = "html")
#install.packages("scales")
library(scales)
#install.packages("tidygenomics")
library(tidygenomics)
```

# Genomic dataset - A few clarifications

Vincenzo Coia has approved my request to use published genomic data for the next assignments, therefore, I want to provide a few clarifications:

* I have tried to introduce some basic explanations about the genomic dataset, but obviously, this is not a genetics course and my objective is to explore and learn how to use R and its packages, not to teach complex notions of cancer genomic analysis. Therefore, I don't expect people to understand what the data and plots represent if they haven't studied in related fields.  

## SOURCE OF DATA: The Cancer Genome Atlas (TCGA)

After trying to complete this homework with the genomic data previously used in [homework 3](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat545-hw3-thibodeau-mylinh), I realized that it was impossible for me to do some of the exercises because I didn't have the right "data.frame" format for it. 

The results shown here are in whole or part based upon data generated by the TCGA Research Network: http://cancergenome.nih.gov/. 

Genomic datasets are available through the cBioportal [here](http://www.cbioportal.org/data_sets.jsp)

# Setting up the data

## Getting the data ready

Let's select some RNAseq (gene expression) cancers datasets from TCGA:

* brca: breast cancer
* gbm: glioblastoma
* pcpg: pheochromocytome-paraganglioma
* sarc: sarcoma
* skcm: skin cutaneous melanoma
* coadread: colorectal adenocarcinoma
* dlbc: diffuse large B-cell lymphoma

I will only select 50 genes (nrows = 100) for each dataset, because there are way to many genes (18000-20000) to keep them all, and also, we will only select 3 samples per cancer type ([ ,c(1,3,4,5)]):
```{r}
brca_rna <- read.table("/Users/mylinh/Documents/published_data/tcga/brca/tcga/data_RNA_Seq_v2_expression_median.txt", header = TRUE, nrows =100)[ ,c(1,3,4,5)]
colnames(brca_rna) <- c("hugo", "brca_s1", "brca_s2", "brca_s3")
brca_rna <- brca_rna[!duplicated(brca_rna$hugo), ]
brca_rna$cancer_type <- "BRCA"
View(brca_rna)

gbm_rna <- read.table("/Users/mylinh/Documents/published_data/tcga/gbm/tcga/data_RNA_Seq_v2_expression_median.txt", header = TRUE, nrows =100)[, c(1,3,4,5)]
colnames(gbm_rna) <- c("hugo", "gbm_s1", "gbm_s2", "gbm_s3")
gbm_rna <- gbm_rna[!duplicated(gbm_rna$hugo), ]
gbm_rna$cancer_type <- "GBM"
#View(gbm_rna)

pcpg_rna <- read.table("/Users/mylinh/Documents/published_data/tcga/pcpg/tcga/data_RNA_Seq_v2_expression_median.txt", header = TRUE, nrows =100)[, c(1,3,4,5)]
colnames(pcpg_rna) <- c("hugo", "pcpg_s1", "pcpg_s2", "pcpg_s3")
pcpg_rna <- pcpg_rna[!duplicated(pcpg_rna$hugo), ]
pcpg_rna$cancer_type <- "PCPG"
#View(pcpg_rna)

sarc_rna <- read.table("/Users/mylinh/Documents/published_data/tcga/sarc/tcga/data_RNA_Seq_v2_expression_median.txt", header = TRUE, nrows =100)[, c(1,3,4,5)]
colnames(sarc_rna) <- c("hugo", "sarc_s1", "sarc_s2", "sarc_s3")
sarc_rna <- sarc_rna[!duplicated(sarc_rna$hugo), ]
sarc_rna$cancer_type <- "SARC"
#View(sarc_rna)

skcm_rna <- read.table("/Users/mylinh/Documents/published_data/tcga/skcm/tcga/data_RNA_Seq_v2_expression_median.txt", header = TRUE, nrows =100)[, c(1,3,4,5)]
colnames(skcm_rna) <- c("hugo", "skcm_s1", "skcm_s2", "skcm_s3")
skcm_rna <- skcm_rna[!duplicated(skcm_rna$hugo), ]
skcm_rna$cancer_type <- "SKCM"
#View(skcm_rna)

coadread_rna <- read.table("/Users/mylinh/Documents/published_data/tcga/coadread/tcga/data_RNA_Seq_v2_expression_median.txt", header = TRUE, nrows =100)[, c(1,3,4,5)]
colnames(coadread_rna) <- c("hugo", "coadread_s1", "coadread_s2", "coadread_s3")
coadread_rna <- coadread_rna[!duplicated(coadread_rna$hugo), ]
coadread_rna$cancer_type <- "COADREAD"
#View(coadread_rna)

dlbc_rna <- read.table("/Users/mylinh/Documents/published_data/tcga/dlbc/tcga/data_RNA_Seq_v2_expression_median.txt", header = TRUE, nrows =100)[, c(1,3,4,5)]
colnames(dlbc_rna) <- c("hugo", "dlbc_s1", "dlbc_s2", "dlbc_s3")
dlbc_rna <- dlbc_rna[!duplicated(dlbc_rna$hugo), ]
dlbc_rna$cancer_type <- "DLBC"
#View(dlbc_rna)
```
Note1. I populated one additional column per dataset with the cancer_type as this will represent my "gapminder$continent" value later on. I tried the technique explained on [computer world here](https://www.computerworld.com/article/2486425/business-intelligence/business-intelligence-4-data-wrangling-tasks-in-r-for-advanced-beginners.html), and although it didn't work out, it gave me the idea of creating a new column "cancer_type" with a single value for the specific cancer.
Note2. I found a way of selecting specific columns in read.table with [this stackoverflow dicussion](https://stackoverflow.com/questions/5788117/only-read-limited-number-of-columns) and [this stackexchange discussion](https://stats.stackexchange.com/questions/16796/reading-only-two-out-of-three-columns-with-read-csv).
Note3. After experiencing issues later on in this homework, I had to remove the hugo genes that had 2 or more values for the same cancer sample. I needed to have only one gene-sample key for the spread() function to workout. I found the information about how to remove duplicate [here](https://stackoverflow.com/questions/13279582/select-only-the-first-rows-for-each-unique-value-of-a-column-in-r)

# Intertwined: General data reshaping and relationship to aggregation AND Join, merge, look up 

*In order to be able to re-shape my data table, I had to complete some exercise out of order, so the order of the homework below looks like:*

* General data reshaping: Activity #1 - GATHER
* Join, merge, look up: Activity #1 - Create the "full_dataset" (equivalent of gapminder)
* General data reshaping: Activity #1 - SPREAD
* Join, merge, look up: Activity #1 - Create a second dataset "clinvar_subset" and manipulate/join the data with the "full_dataset"
* 

## General data reshaping: Activity #1 - Make your own cheetsheet

*Make you own cheatsheet similar to Tyler Rinkerâ€™s [minimal guide to tidyr](https://github.com/trinker/tidyr_in_a_nutshell).*

I also like this [Data Wrangling with dplyr and tidyr cheat sheet](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf).

I found a few tutorials that I liked as well:

* University of Virginia: [A tidyr Tutorial](http://data.library.virginia.edu/a-tidyr-tutorial/)
* RPubs of RStudio: [Data Processing with dplyr & tidyr](https://rpubs.com/bradleyboehmke/data_wrangling) by Brad Boehmke
* Tidyverse website: [tidyr article](http://tidyr.tidyverse.org/)

### GATHER: From horizontal (wide) to vertical (longer)

According the the Help of RStudio, the gather() function is a way to take columns that are not variables and collapse multiple columns to key-value pairs. 

For my type of data, I think gather() will be useful when I realize that a group of columns actually represent the same variable and therefore, they should be "stacked on top of each other" () rather than in adjacent columns. As Hadley Wickham said in his [Introducing tidyr blog](https://blog.rstudio.com/2014/07/22/introducing-tidyr/),the most important for tidy data is that each column is a variable and each row is an observation.

Let's take a look at the brca_rna data:
```{r}
brca_rna %>%
  kable("html") %>% kable_styling()
```

The TCGA RNAseq datasets are perfect for this exercise, because the columns in each dataset represent a different sample, and therefore, they are actually different obversations. I will start by making new "gathered" data.frame.
```{r}
brca_rna_gather <- brca_rna %>%
  gather(key = sample, value = RPKM, brca_s1, brca_s2, brca_s3)
#View(brca_rna_gather)
gbm_rna_gather <- gbm_rna %>%
  gather(key=sample, value = RPKM, gbm_s1, gbm_s2, gbm_s3)
pcpg_rna_gather <- pcpg_rna %>%
  gather(key = sample, value = RPKM, pcpg_s1, pcpg_s2, pcpg_s3)
sarc_rna_gather <- sarc_rna %>%
  gather(key=sample, value = RPKM, sarc_s1, sarc_s2, sarc_s3)
skcm_rna_gather <- skcm_rna %>%
  gather(key = sample, value = RPKM, skcm_s1, skcm_s2, skcm_s3)
coadread_rna_gather <- coadread_rna %>%
  gather(key = sample, value = RPKM, coadread_s1, coadread_s2, coadread_s3)
dlbc_rna_gather <- dlbc_rna %>%
  gather(key =sample, value = RPKM, dlbc_s1, dlbc_s2, dlbc_s3)
```

And now, if we take the same brca data, but the after using the gather() function:
```{r}
brca_rna_gather %>%
  kable("html") %>% kable_styling()
```

### SPREAD: From vertical (longer) to horizontal (wide)

In order to revert back to the original format of the data, I can use the same key:value pairing in the spread() function. For example, the following code's output is the original data format since gather() and spread() are mirror functions from each other.

```{r}
brca_rna %>%
  gather(key = sample, value = RPKM, brca_s1, brca_s2, brca_s3) %>%
  spread(key=sample, value=RPKM) %>%
  kable("html") %>% kable_styling()
```
*So it turns out that the spread() function does not like the fact that some gene names are duplicated for the sample sample initially, as it returned the message "Error: Duplicate identifiers for rows (2, 3), (52, 53), (102, 103)". So I have decided to change apply some extra functions to filter the duplicate when I read the TCGA cancer datasets files at the beginning of the homework.*

## Join, merge, look up: Activity #1

For the next step, I will have to skip ahead in the tasks a little and use some joining funcitons to compound all these TCGA datasets into one "gapminder-like" dataset.
```{r}
list_datasets <- list(brca_rna_gather, gbm_rna_gather, pcpg_rna_gather, sarc_rna_gather, skcm_rna_gather, coadread_rna_gather, dlbc_rna_gather)
# View(list_genomic_datasets_v3)
full_dataset <- list_datasets %>%
  Reduce(function(x, y) full_join(x, y), .)
```
Note. I won't lie, I needed to read a lot in order to figure the chunk of code above:

* 

## General data reshaping: Activity #2 - Make a tibble with one row per year and columns for life expectancy for two or more countries.

I will make a tibble with one row per gene (year) and columns for RPKM values (life expectancy) for two of more cancer samples (countries) in the BRCA group (continent).

```{r}
full_dataset %>%
  filter(cancer_type=="BRCA") %>%
  spread(key=sample, value =RPKM) %>%
  kable("html") %>% kable_styling()
```


## General data reshaping: Activity #3

Compute some measure of RPKM (lifeExp) (mean? median? min? max?) for all possible combinations of cancer group (continent) and hugo (year). 

```{r}
full_dataset %>%
  group_by(hugo, cancer_type) %>%
  summarize(mean_cancer=signif(mean(RPKM),2), median_cancer = signif(median(RPKM), 2), min_cancer=signif(min(RPKM), 2), max_cancer=signif(max(RPKM), 2)) %>%
  kable("html") %>% kable_styling()
```

Reshape that to have one row per hugo (year) and one variable for each cancer group (continent). 

This sounds like spread() and gather() would be useful here. So let's start by doing this exercise with the mean_cancer variable.
```{r}
full_dataset %>%
  group_by(hugo, cancer_type) %>%
  summarize(mean_cancer= signif(mean(RPKM), 2)) %>%
  gather(key=stat_category, value= the_number , mean_cancer) %>%
  spread(key=cancer_type, value=the_number) %>%
  kable("html") %>% kable_styling()
```

*Wow ... that took a long time, so let me reason through the process for one second:*

* I grouped by hugo gene (year) and cancer_type (continent)
* Got my summary stat mean_cancer 
* Used gather() with a key = stat_category and a value = the_number and mean_cancer
- Key = new variable created = stat_category
- Value = column name representing the value = the_number, which is a temporary variable I created to use in the spread() function after
- Names of columns to gather = mean_cancer
* Used spread() with a key = cancer_type and a value=the_number
- Which means that the cancer_type now became columns of their own
- And I indicated the value = the_number, which means the mean_cancer value in this case.

And now, we can extend it to more summary statistics:
```{r}
full_dataset %>%
  group_by(hugo, cancer_type) %>%
  summarize(mean_cancer= signif(mean(RPKM), 2), median_cancer = signif(median(RPKM), 2), min_cancer=signif(min(RPKM), 2), max_cancer=signif(max(RPKM), 2)) %>%
  gather(key=stat_category, value= the_number , mean_cancer, median_cancer, min_cancer, max_cancer) %>%
  spread(key=cancer_type, value=the_number) %>%
  kable("html") %>% kable_styling()
```

Or the other way around: one row per cancer group (continent) and one variable hugo gene (year). Let's just do all the summary statistics on this one.
```{r}
full_dataset %>%
  group_by(cancer_type, hugo) %>%
  summarize(mean_cancer= signif(mean(RPKM), 2), median_cancer = signif(median(RPKM), 2), min_cancer=signif(min(RPKM), 2), max_cancer=signif(max(RPKM), 2)) %>%
  gather(key=stat_category, value= the_number , mean_cancer, median_cancer, min_cancer, max_cancer) %>%
  spread(key=hugo, value=the_number) %>%
  kable("html") %>% kable_styling()
```

***

Trying to read clinvar data.
```{r}
clinvar_subset <- read.table("/Users/mylinh/Documents/published_data/subset_clinvar.txt", sep = "\t", header=TRUE)
View(clinvar_subset)
```





Second, as you can see, the 

# So now we can do the first part of the homework: 
***


