---
title: "stat545-hw05-thibodeau-mylinh"
author: "My Linh Thibodeau"
date: '2017-10-14'
output: github_document
---

```{r}
suppressPackageStartupMessages(library(tidyverse))
knitr::opts_chunk$set(fig.width=12, fig.height=9)
library(knitr)
library(kableExtra)
options(knitr.table.format = "html")
library(reshape2)
#install.packages("tidygenomics") # Not used for now, but perhaps in another homework
#library(tidygenomics) # Not used for now, but perhaps in another homework
```

# Genomic datasets - A few clarifications

Vincenzo Coia has approved my request to use published genomic data for the next assignments, therefore, I want to provide a few clarifications:

* I have tried to introduce some basic explanations about the genomic dataset, but as my objective is to explore and learn how to use R and its packages, not to teach cancer genomic analysis, I am sorry in advance if you are not sure what the data and plots represent, but focus on the data manipulation behind and not what the data represents, and I hope you will still enjoy reading my homework !
* If you have any questions, please don't hesitate to let me know :) 
* Given the size of the data, I only display a few rows per table to keep it readable.
* If you are not sure of the corresponding variables (or equivalences) between gapminder and my genomic datasets, you can always go back to the README.md file [here](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat545-hw4-thibodeau-mylinh)
* RPKM (Reads Per Kilobase of transcript per Million mapped reads) is a measure of gene expression obtained from RNAseq data.

## SOURCES OF DATA: The Cancer Genome Atlas (TCGA) and ClinVar 

After trying to complete this homework with the genomic data previously used in [homework 3](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat545-hw3-thibodeau-mylinh), I realized that it was impossible for me to do some of the exercises because I didn't have the right "data.frame" format for it. 

**The Cancer Genome Atlas (TCGA)**    
The results shown here are in whole or part based upon data generated by the TCGA Research Network: http://cancergenome.nih.gov/

Genomic datasets are available through the cBioportal [here](http://www.cbioportal.org/data_sets.jsp)

**ClinVar**  
The results shown here are in whole or part based upon the data generated by the NCBI ClinVar data portal: https://www.ncbi.nlm.nih.gov/clinvar/

The ClinVar genomic variant dataset is publicly available [here](ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/). If you would like more information on the type of files available, please read this [primer](https://www.ncbi.nlm.nih.gov/clinvar/docs/ftp_primer/) on the topic.

