---
title: "hw10-unwrapped-api"
author: "My Linh Thibodeau"
date: '2017-11-29'
output: github_document
---

```{r}
suppressMessages(suppressWarnings(library(XML)))
suppressMessages(suppressWarnings(library(xml2)))
suppressMessages(suppressWarnings(library(dplyr)))
knitr::opts_chunk$set(fig.width=12, fig.height=9)
suppressMessages(suppressWarnings(library(knitr)))
suppressMessages(suppressWarnings(library(kableExtra)))
options(knitr.table.format = "html")
suppressMessages(suppressWarnings(library(tidyverse)))
suppressMessages(suppressWarnings(library(purrr)))
suppressMessages(suppressWarnings(library(stringr)))
suppressMessages(suppressWarnings(library(forcats)))
suppressMessages(suppressWarnings(library(httr)))
suppressMessages(suppressWarnings(library(jsonlite)))
suppressMessages(suppressWarnings(library(glue)))
suppressMessages(suppressWarnings(library(methods)))
```


# Introduction

In this homework, I will be working with the Online Mendelian Inheritance of Man - [OMIM](https://omim.org/) data as well as previously used merged data from [DECIPHER](http://decipher.sanger.ac.uk), [Orphanet](http://www.orpha.net) and [HGNC](https://www.genenames.org/cgi-bin/statistics) (see references and [homework 8](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw8-thibodeau-mylinh) for more information)

Extensive information on the OMIM API is available [here](https://omim.org/help/api), but I will provide a few examples of query and functions to query the API.

**IMPORTANT:** This work is adapted from Dave Tang's tutorial on Getting started with the OMIM api [here](https://davetang.org/muse/2015/03/17/getting-started-with-the-omim-api/).

**IMPORTANT:** An alternative method has been used to retrieve the OMIM API data: I mainly used adapted functions instead of `GET()` from the library httr.

**IMPORTANT:** I spent quite a bit of time trying to make some output files that you guys could see while respecting the protected data restriction of the OMIM API. Please install the packages XML and xml2 in order to be able to take a better look at the OMIM API xml output examples in the [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample), and otherwise, I have put some data samples in the homework, but the aesthetic of it is not great. 

# Homework 10 requirements

Here is the breakdown of requirements and corresponding sections:

* Get data from API - Section A (part 1, 2, 3, 4, 5) and Section D (part 1) 
* Clean and tidy into dataframes - Section B (part 2) and Section C and Section D (part 2)
* Save output to files
	- protected data has been saved under the data folder - not public (e.g. mim2gene_and_inheritance_DDG2P_hgnc_orphadata.tsv)
	- sample data for reviewers saved under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample)
* Use httr and `GET()` - Section F contains some examples of API query using httr. A new function using httr and glue is defined.

## SECTION A - Writing functions to work with OMIM API

### Part 1 - Get the OMIM API key "privately"

I have stored privately my OMIM API key and I will set the key for later use in building queries.
```{r}
key <- getOption("private_omim_api_key")
my_key <- paste0("apiKey=", key)
```


### Part 2 - Function to build URL and get XML

The general format of query (defaul XML) is `http://api.omim.org/api/entry?mimNumber=100100&apiKey=key&include=text&include=geneMap`. This link does not work because my API key is private, but it gives you a general idea of the format.

```{r}
get_omim_to_xml <- function(
	omim_id = 277700,
	text = FALSE, #Includes the text field sections with the entry.
  	existflags = FALSE, #Include the 'exists' flags with the entry (clinical synopsis, allelic variant, gene map & phenotype map).
  	allelicVariantList = FALSE, #Includes the allelic variant list with the entry.
  	clinicalSynopsis = FALSE, #Include the clinical synopsis with the entry.
  	seeAlso = FALSE, #Includes the 'see also' field with the entry.
  	referenceList = FALSE, #Include the reference list with the entry.
  	geneMap = FALSE, #Include the gene map/phenotype map data with the entry.
  	externalLinks = FALSE, #Include the external links with the entry.
  	contributors = FALSE, #Includes the 'contributors' field with the entry.
  	creationDate = FALSE, #Includes the 'creation date' field with the entry.
  	editHistory = FALSE, #Includes the 'edit history' field with the entry.
  	dates = FALSE, #Include the dates data with the entry.
  	all = FALSE #Include the above data with the entry.
){
	# Obtain the arguments the functions calls to build an URL and get XML
	list_arguments <- as.list(match.call())
  	my_mim   <- paste('mimNumber=', omim_id, sep='')
  	my_link  <- 'http://api.omim.org/api/entry?'
  	my_query <- paste(my_link, my_mim, my_key, sep = "&")
  	
  	#loop through all the arguments
  	for (elem in names(list_arguments)){
  		#skip the omid_id and blank argument
  		if(!elem %in% '' && !elem %in% 'omim_id'){
  			 my_include <- paste('&', 'include=', elem, sep='')
  			 my_query <- paste(my_query, my_include, sep='')
  			 }
  		}
xmlParse(my_query)
}
```

#### Peak at data output

I have set the default OMIM id to 277700, which corresponds to Werner syndrome.

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# get_omim_xml_werner_example1 <- get_omim_to_xml()
# class(get_omim_xml_werner_example1)
# saveXML(get_omim_xml_werner_example1, "data_sample/get_omim_xml_werner_example1.xml")
read_xml("data_sample/get_omim_xml_werner_example1.xml") %>% 
	xmlParse() 
```

Let's peak at the structure of the geneMap and the information it contains.

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# get_omim_xml_werner_example2 <- get_omim_to_xml(geneMap = TRUE)
# saveXML(get_omim_xml_werner_example2, "data_sample/get_omim_xml_werner_example2.xml")
read_xml("data_sample/get_omim_xml_werner_example2.xml") %>% 
	xmlParse() 
```

### Part 3 - Function get_gene (example of "nested" function to parse XML)

Code adapted from Dave Tang [here](https://github.com/davetang/romim).
```{r}
get_gene <- function(my_xml){
  my_gene_node <- getNodeSet(my_xml, path = "/omim/entryList/entry/phenotypeMapList/phenotypeMap/geneSymbols")
  xmlSApply(my_gene_node, xmlValue)
}
```

And then we can retrieve the gene corresponding to a specific OMIM id (default is Werner syndrome).

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# get_gene_werner_example3 <- get_gene(get_omim_to_xml(geneMap = TRUE)) %>% as.data.frame()
# write_tsv(get_gene_werner_example3, "data_sample/get_gene_werner_example3.tsv")
read.table("data_sample/get_gene_werner_example3.tsv", sep = "\t", header = TRUE) %>% kable()
```

### Part 4A - Function get_title (example of XML parsing)

Now, we need a function to parse the XML and retrieve the disorder **titles** corresponding to the OMIM id. 
```{r}
get_title <- function(xml =  277700){
  xml_list <- xmlToList(xml)
  return(xml_list$entryList$entry$titles$preferredTitle)
}
```

#### Peak at the data output

Using the default option (Werner syndrome), we obtained the preferredTitle of this disorder. 

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# get_title_werner_example4 <- get_title(get_omim_to_xml()) %>% as.data.frame()
# write_tsv(get_title_werner_example4, "data_sample/get_title_werner_example4.tsv")
read.table("data_sample/get_title_werner_example4.tsv", sep = "\t", header = TRUE) %>% kable()
```

### Part 4B - Write a get titles function that accepts a list of OMIM ids

```{r}
get_titles_from_list <- function(
	user_list = list(187500, 214800, 188400, 217095) # Set default as diseases with tetralogy of fallot (very frequent or frequent)
	){
	if (is.list(user_list)){
		my_list_titles <- lapply(lapply(user_list, get_omim_to_xml), get_title)
		df <- data.frame(id=unlist(user_list),
                 title=unlist(my_list_titles))
		return(df)
	}
	else
		return("You have to provide an R Object typeof list.")
}
```

#### Peak at the data output

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# get_titles_from_list_tetralogy_fallot <- get_titles_from_list() 
# write_tsv(get_titles_from_list_tetralogy_fallot, "data_sample/get_titles_from_list_tetralogy_fallot.tsv")
read.table("data_sample/get_titles_from_list_tetralogy_fallot.tsv", sep = "\t", header = TRUE) %>% kable()
```


### Part 5 - Get multiple items from OMIM API

So far, we got the titles (diseases names) from the API, but we can also see what we could do with the clinical synopsis. Let's start by peaking at the structure of xml for clinical synopsis data.

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# get_omim_xml_werner_example5 <- get_omim_to_xml(clinicalSynopsis = TRUE)
# saveXML(get_omim_xml_werner_example5, "data_sample/get_omim_xml_werner_example5.xml")
read_xml("data_sample/get_omim_xml_werner_example5.xml") %>% 
	xmlParse() 
```


```{r}
get_clinical_synopsis_as_list <- function(xml){
	xml_list <- xmlToList(xml)
	#my_clinical_node <- getNodeSet(my_xml, path = "/omim/entryList/entry/clinicalSynopsisList/clinicalSynopsis")
  	#xmlSApply(my_clinical_node, xmlValue)
	my_query <- xml_list$entryList$entry$clinicalSynopsis
	#my_query_parse <- xml_list()
  	return(my_query)
}
```

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# clinical_synopsis_werner_list <- get_clinical_synopsis_as_list(get_omim_to_xml(clinicalSynopsis = TRUE))
# capture.output(clinical_synopsis_werner_list, file = "data_sample/clinical_synopsis_werner_list.txt")
read.table("data_sample/clinical_synopsis_werner_list.txt", sep="\t") %>% kable()
```
Note. I know the output is not visually pleasing, but I am just trying to show that the API query worked and to prove to you that I was able to obtain the desired web data outpub. 

## SECTION B - Merging data from mim2gene dataset with OMIM API output from functions

### Part 1 - Get the mim2gene data

Complementary to getting access to the OMIM API, OMIM allowed me to download the mim2gene.txt data, which lists OMIM ids and the corresponding OMIM entry type, Entrez gene id, HGNC is and Ensembl id.

```{r}
mim2gene_data <- read.table("data/mim2gene.txt", header=FALSE, sep="\t")
colnames(mim2gene_data) <- c("id", "mim_entry_type", "entrez_gene_id", "hgnc", "ensembl_gene_id")
```

#### Peak at the data output

```{r}
head(mim2gene_data) %>% kable()
```


### Part 2 - Merge with OMIM API data

The mim2gene_data is not really useful in this state, but if we leverage the data from the OMIM API, we will be able to expand this table and make useful genotype-phenotype correlations.

```
system.time(mim2gene_get_mim_api_titles <- lapply(lapply(mim2gene_data$id, get_omim_to_xml), get_title))
```
	user   system  elapsed 
  99.193   13.788 5675.498 

*So it did take about 3h, as mentioned by David Tang.* 

Transform the largest list obtained into a dataframe.
```
df <- data.frame(id=mim2gene_data$id,
                 title=unlist(mim2gene_get_mim_api_titles))
```

Let's merge this list of diseases titles with the previous mim2gene_data table. 
```
mim2gene_data_expanded_titles <- merge(x=mim2gene_data, y=df, by='id')
dim(mim2gene_data_expanded_titles)
```

Let's save this new dataset.
```
write_tsv(mim2gene_data_expanded_titles, "data/mim2gene_data_expanded_titles.tsv")
```

*Since I haven't re-queried the database for all the diseases titles (which takes 3h), I have to read the table I generated earlier.*
```{r}
mim2gene_data_expanded_titles <- read.table("data/mim2gene_data_expanded_titles.tsv", sep = "\t", header = TRUE)
```

*And let's store the diseases OMIM ids in a variable for later use.*

```{r}
omim_db_ids <- mim2gene_data_expanded_titles$id %>% as.data.frame()
colnames(omim_db_ids) <- c("disease.mim")
head(omim_db_ids)
```


So now we have a much more informative dataframe, which contains both the OMIM ids and the diseases names (**titles**).
```{r}
# dim(mim2gene_data_expanded_titles)
mim2gene_data_expanded_titles %>% head() %>% kable()
```


## SECTION C - BUILDING ON PREVIOUS STAT545/547 HOMEWORK 

Data manipulation (joining, cleaning and tidying) has been a necessary (and often painful, as explained in most repository README) step in most of my previous STAT545/547 homework and I encountered common challenges. If at this point, you are not convinced that I can obtain from the web and manipulate messy data, you have to checkout my previous homeworks. Here are some examples:

- [homework 3](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat545-hw3-thibodeau-mylinh) - [tables and plots](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/blob/master/stat545-hw3-thibodeau-mylinh/stat545-hw03-thibodeau-mylinh.md) - Challenge with duplicated factors (genes names)
- [homework 4](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat545-hw4-thibodeau-mylinh) - [long version](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/blob/master/stat545-hw4-thibodeau-mylinh/long-version-stat545-hw04-thibodeau-mylinh.md) - Challenge with data structure, as I realized midway that I could not perform some of the required analyses due to the characteristics of my dataset, and I had to restart with a new dataset (The Cancer Genome Atlas).
- [homework 5](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat545-hw5-thibodeau-mylinh) - [tables and plots](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/blob/master/stat545-hw5-thibodeau-mylinh/stat545-hw05-thibodeau-mylinh.md) - Challenge with `read_table2()`, and subsequent discovery that some reading functions do no deal with white spaces well. 
- [homework 6](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/blob/master/stat547-hw6-thibodeau-mylinh/stat547-hw06-thibodeau-mylinh.Rmd) - [gene ontology and Genomic Data Commons](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/blob/master/stat547-hw6-thibodeau-mylinh/stat547-hw06-thibodeau-mylinh.Rmd) - Challenges dealing with a new file format (".obo") and new data structure/organization of gene ontology, and challenges dealing with nested list of Genomic Data Commons (cancer data), and the whole SSL certificate issue.
- [homework 7](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw7-thibodeau-mylinh) - [mut_sig_tables.R](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/blob/master/stat547-hw7-thibodeau-mylinh/mut_sig_tables.R) - Multiple data manipulation steps with `gather()` function of tidyr, and challenges in trying to join them with mutational signatures.
- [homework 8](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw8-thibodeau-mylinh) - [dataframe manipulation](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/blob/master/stat547-hw8-thibodeau-mylinh/genomic-disorders-exploration-data-manipulation.Rmd) - There were so many steps of data manipulation required that I decided to make and Rmd file and explain the process as part of building my ShinyApp - Genomic Disorders Exploration.
- [homework 9](https://github.com/mylinhthibodeau/powers3) - Challenge with the superpowers dataset, which contains non ASCII characters (still under work at this time).


Therefore, instead of reinventing the wheel, I will build on some of the previous work accomplished ! Specifically, I will use the data from my [homework 8](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw8-thibodeau-mylinh) Shiny App (partial data available [here](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw8-thibodeau-mylinh/data))

#### Part 1 - Read and peak at merged dataset (DECIPHER, HGNC and Orphanet)

Now, let's read the previously generated merged data from DECIPHER, HGNC and Orphanet as well as the list of omim disease ids from this merged dataset. 
```{r}
DDG2P_hgnc_orphadata <- read.table("data/DDG2P_hgnc_orphadata.tsv", header = TRUE, sep = "\t")
DDG2P_hgnc_orphadata_list_disease_mim <- read.table("data/DDG2P_orphadata_list_disease.mim.txt", header = TRUE, sep = "\t")
```

Please refer to homework 8 for additional details on these datasets, but here is a little peak at the data.
```{r}
DDG2P_hgnc_orphadata %>% head() %>% kable()
```

And a peak at the diseases OMIM ids.

```{r}
# dim(DDG2P_hgnc_orphadata_list_disease_mim)
# class(DDG2P_hgnc_orphadata_list_disease_mim)
DDG2P_hgnc_orphadata_list_disease_mim %>% head() %>% kable()
```

Now, let's find the OMIM ids common to OMIM database and orphadata

```{r}
common_mim_ids <- inner_join(DDG2P_hgnc_orphadata_list_disease_mim, omim_db_ids, by = c( "disease.mim" = "disease.mim"))
dim(common_mim_ids)
```

Let's make a list of these omim diseases ids of interest.
```{r}
my_omim_list <- common_mim_ids$disease.mim %>% as.list()
```


## SECTION D - TARGETED API QUERIES

Since we saw that an OMIM database-wide query can take up to 3h, we will only make targeted data queries on a subset of the OMIM data now.

### Part 1 - Function get_inheritance (example of XML parsing)

Now, we need a function to parse the XML and retrieve the disorder **inheritance** corresponding to the OMIM id. 
```{r}
get_inheritance <- function(my_xml){
    my_inheritance_node <- NULL
	my_inheritance_node <- getNodeSet(my_xml, path = "/omim/entryList/entry/phenotypeMapList/phenotypeMap/phenotypeInheritance")
  my_inheritance_result <- xmlSApply(my_inheritance_node, xmlValue)
  
  if (is.null(my_inheritance_node)){
  	
  }
  return(my_inheritance_result)
}

```

Let's peak at the output for mim.disease 303600 (Coffin-Lowry syndrome).

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# get_inheritance_coffin_lowry_example6 <- get_inheritance(get_omim_to_xml(303600, geneMap=TRUE))
# capture.output(get_inheritance_coffin_lowry_example6, file = "data_sample/get_inheritance_coffin_lowry_example6.txt")
read.table("data_sample/get_inheritance_coffin_lowry_example6.txt", sep = "\t") %>% kable()
```

Let's make a function that accepts a list of OMIM ids and provides the inheritance for these 918 OMIM ids. The first step is to make a function that will determine if the data is already available or not.

```{r}
get_inheritance_from_list <- function(my_file, user_list){
	inheritance_list <- try(read_tsv(my_file, col_names = FALSE))
	if (!inherits(inheritance_list, "try-error")) {
  		read_tsv(my_file)
  		return("The file is present")
		}
	else
		inheritance_list <- list()
		if (is.list(user_list)){
			for (i in 1:length(user_list)){
				inheritance <- get_inheritance(get_omim_to_xml(user_list[i], geneMap=TRUE))
				
				if (length(inheritance) >1){
					inheritance <- paste(inheritance, collapse = "/")
				}
				if (is.list(inheritance) || inheritance == ""){ inheritance_list[i] <- "0" # Missing values are replaced by the "0" character
				} 
				else
					{ inheritance_list[i] <- as.character(inheritance)}
			}
		}
	if (length(inheritance_list) !=0){
		inheritance_df <- inheritance_list %>% as.data.frame()
		write.table(inheritance_df, "data_sample/get_inheritance_from_my_omim_list.tsv", sep = "\n", col.names = FALSE)
		print("The inheritance file was not available, an OMIM API query has been performed and a file written to the folder data_sample/ .")
	}
	else
		print("The inheritance file was available and data has been read.")
	return(inheritance_list)
	}
```

Let's test the function!

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# my_file <- "stat547-hw10-thibodeau-mylinh/data_sample/get_inheritance_from_my_omim_list.tsv"
# my_omim_list <- common_mim_ids$disease.mim %>% as.list() # we already defined this list in a previous section
# get_inheritance_from_my_omim_list <- get_inheritance_from_list(my_file, my_omim_list)
#View(get_inheritance_from_my_omim_list)
get_inheritance_from_my_omim_list2 <- read.table("data_sample/get_inheritance_from_my_omim_list.tsv", sep = "\t", header = FALSE) %>% as.list()
get_inheritance_from_my_omim_list2 <- get_inheritance_from_my_omim_list2$V1
```
(message from function pasted below as an example)
`[1] "The inheritance file was not available, an OMIM API query has been performed and a file written to the folder data_sample/ ."`

#### Peak at the data output

Let's look at what kind of data we are dealing with here (and make sure we have the same number of items we started with).
```{r}
length(get_inheritance_from_my_omim_list2) # For some reason, there is an added row with the number 1, I will remove it
get_inheritance_from_my_omim_list2 <- get_inheritance_from_my_omim_list2[2:919]
class(get_inheritance_from_my_omim_list2)
typeof(get_inheritance_from_my_omim_list2)
get_inheritance_from_my_omim_list2 %>% head(10) %>% kable()
```
Both the inheritance list and the omim ids list have 918 items, perfect ! Let's create a dataframe with them, so that we can join this data to the bigger dataset (DECIPHER, HGNC, Orphadata from Orphanet, mim2gene).

### Part 2 - Make some dataframes !

```{r}
inheritance_mim_df <- data_frame(as.vector(my_omim_list), as.vector(get_inheritance_from_my_omim_list2))
colnames(inheritance_mim_df) <- c("disease.mim", "inheritance")
inheritance_mim_df %>% head() %>% kable()
```

Merge with the previously made dataframe of mim2gene_data_expanded_titles.
```{r}
# In order to merge the data, we need to change disease mim ids to character data (joining character and integer does not work).
mim2gene_data_expanded_titles$id <- as.character(mim2gene_data_expanded_titles$id)
#d1 <- apply(mim2gene_data_expanded_titles,2,function(x)gsub('\\s+', '',x))
inheritance_mim_df$disease.mim <-  as.character(inheritance_mim_df$disease.mim)
class(inheritance_mim_df$disease.mim)
length(inheritance_mim_df$disease.mim)
mim2gene_and_inheritance <- inner_join(inheritance_mim_df, mim2gene_data_expanded_titles, by = c("disease.mim" = "id" ))
# dim(mim2gene_and_inheritance)
# View(mim2gene_and_inheritance)
```

Find the common overlap between the DDG2P_hgnc_orphadata and the mim2gene_and_inheritance datasets (I am purposefully reducing the dimensions of the datasets here to facilitate data overview).

```{r}
DDG2P_hgnc_orphadata_subset <- DDG2P_hgnc_orphadata %>%
	group_by(disease.mim) %>%
	slice(1L)
DDG2P_hgnc_orphadata_subset$disease.mim <- as.character(DDG2P_hgnc_orphadata_subset$disease.mim)
mim2gene_and_inheritance_DDG2P_hgnc_orphadata <- DDG2P_hgnc_orphadata_subset %>% 
	inner_join(mim2gene_and_inheritance, by = c("disease.mim" = "disease.mim"))
# View(mim2gene_and_inheritance_DDG2P_hgnc_orphadata)
#mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance %>% unique() 
```

Some of the inheritance labels got duplicated, so I will try to clean the data and used abbreviations here. 
```{r}
# mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance %>% unique() 
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "Autosomal dominant", "AD")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "Autosomal recessive", "AR")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "AD/AD", "AD")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "AR/AR", "AR")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "AD/AD", "AD")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "AR/AR", "AR")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "AD/AD", "AD")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "/AD", "AD")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "0" , "/") # all unknown values will be replaced by a "/" symbol
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- str_replace_all(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance, "/Digenic recessive" , "")
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance %>% unique() 
```

#### Peak at the data output

```{r}
mim2gene_and_inheritance_DDG2P_hgnc_orphadata %>%
	head() %>%
	kable()
```

#### Write dataset to a file

```{r}
write_tsv(mim2gene_and_inheritance_DDG2P_hgnc_orphadata , "data/mim2gene_and_inheritance_DDG2P_hgnc_orphadata.tsv")
```


### SECTION E - DATA EXPLORATION - ALL TOGETHER

Please note that data exploration for the API-related dataframe has been performed in the relevant sections above: refer to tables sample provided by the `head()` function.

```{r}
summary(mim2gene_and_inheritance_DDG2P_hgnc_orphadata) %>%
	kable()
mim2gene_and_inheritance_DDG2P_hgnc_orphadata %>% 
	group_by(frequency) %>%
	summarize(total_features = n()) %>%
	arrange(desc(total_features)) %>%
	kable()
``` 
Note. Most diseases in the list do not have phenotypic feature frequency specified.

```{r}
mim2gene_and_inheritance_DDG2P_hgnc_orphadata %>% 
	group_by(inheritance) %>%
	summarize(total_diseases = n()) %>%
	arrange(desc(total_diseases))
```
Note. The most frequent mode of disease inheritance in this subset is autosomal dominant (AD = 274) and autosomal recessive (AR = 255), followed by not available (/ =99). 

Digenic recessive, Somatic mosaicism and X-linked recessive are rare modes of inheritance, I wonder which diseases it concerns.
```{r}
mim2gene_and_inheritance_DDG2P_hgnc_orphadata %>%
	ungroup() %>%
 	filter(grepl("Digenic|Somatic|X-linked recessive", inheritance)) %>%
	select(disorder_name, gene.symbol, inheritance) %>% 
	kable()
```


### SECTION E - PLOTS

```{r}
mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance <- as.factor(mim2gene_and_inheritance_DDG2P_hgnc_orphadata$inheritance) # Change inheritance into a factor in order to use forcats to reorder the bars according to frequency
mim2gene_and_inheritance_DDG2P_hgnc_orphadata %>%
	ungroup() %>%
	mutate(inheritance = inheritance %>% fct_infreq()) %>%
			unique() %>%
			group_by(disease.mim) %>%
			ggplot(aes(inheritance)) +
			geom_bar(aes(fill=inheritance)) +
			ggtitle("Inheritance mode of diseases of merged data") +
			theme_bw() + theme(
				plot.title= element_text(color = "grey44", size=18, face="bold"),
				axis.text.x = element_text(angle = 45, hjust = 1))
```

Let's take a look at the types and distribution of OMIM entry type (mim_entry_type) we obtained from the OMIM API query above. 
```{r}
mim2gene_and_inheritance_DDG2P_hgnc_orphadata %>%
  ggplot(aes(x=factor(1), fill=mim_entry_type)) + 
	geom_bar(width =1) + coord_polar("y")
```
Note. We mainly obtained phenotype data and this is not suprising considering that I have obtained this dataset from overlapping the OMIM API data with the merged DECIPHER/HGNC/Orphanet-Orphadata data, the latter being mainly focused on phenotype.

Now, let's look at the phenotype_name data.

```{r}
mim2gene_and_inheritance_DDG2P_hgnc_orphadata %>%
		ungroup() %>%
			filter(phenotype_name != "") %>%
			filter(!is.na(phenotype_name)) %>%
			ggplot(aes(x=factor(1), fill=phenotype_name)) +
			geom_bar(width =1, show.legend = FALSE) +
			coord_polar("y") +
			ggtitle("Phenotypic features - Distribution") +
			theme_bw() + theme(
				plot.title= element_text(color = "grey44", size=18, face="bold"),
				axis.title.x=element_blank(),
				axis.text.x = element_blank())
```
Note. Most features are only mentioned once or twice, so this looks like a color wheel to pick paint colors of my next apartment, but I thought it would still be interesting to have a visual. Let's take a look at the most frequent phenotypes next.

```{r}
mim2gene_and_inheritance_DDG2P_hgnc_orphadata %>%
	ungroup() %>%
	filter(phenotype_name != "") %>%
	filter(!is.na(phenotype_name)) %>%
	group_by(phenotype_name) %>%
	summarize(phenotype_count = n()) %>%
	arrange(desc(phenotype_count)) %>%
	head()
```
Note. Cryptorchidism is a frequent phenotype of the merged dataset we have and it is present in 12 diseases. It is not surprising given the fact that up to 4% of healthy baby boys present cryptorchidism at birth [Virtanen and Topari, 2007](https://academic.oup.com/humupd/article/14/1/49/823056).


## SECTION F - using httr and GET()

As mentioned previously, I used an alternative method to query the OMIM API database, but just in case the homework requirements are strict, I thought I should show that I can also use httr and `GET()` in the setting of API query. 

Let's take another disease for a change (Caffey disease, mim.disease = 114000): for the clinical synopsis, the url is "https://api.omim.org/api/clinicalSynopsis?mimNumber=114000&format=xml".

Therefore, we could define a function that query the API according to the mimNumber.
```{r}
get_clinical_synopsis_httr <- function(mim.disease){
	query_string <- glue("https://api.omim.org/api/clinicalSynopsis?mimNumber={mim.disease}&format=xml&{my_key}")
	clinical_synopsis_result <- GET(query_string)
	clinical_synopsis_content <- content(clinical_synopsis_result)
	clinical_synopsis_content_parse <- XML::xmlParse(clinical_synopsis_content)
	return(clinical_synopsis_content_parse)
}
```

Let's look at the information available for Caffey disease. 

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# clinical_synopsis_caffey <- get_clinical_synopsis_httr(114000)
# saveXML(clinical_synopsis_caffey, "data_sample/clinical_synopsis_caffey_example1.xml")
read_xml("data_sample/clinical_synopsis_caffey_example1.xml") %>% 
	xmlParse() 
```

Maybe we want a function that looks at a lot more things than the clinical synopsis tree format.
```{r}
get_lots_of_info <- function(mim.disease){
	query_string = glue("https://api.omim.org/api/entry?mimNumber={mim.disease}&include=text&format=xml&{my_key}")
	lots_info_result <- GET(query_string)
	lots_info_content <- content(lots_info_result)
	lots_info_content_parse <- xmlParse(lots_info_content)
	return(lots_info_content_parse)
}
```

Let's take a look at some Caffey disease information !

**IMPORTANT** The xml functions can not be knitr to github document as they are using my private key. As a replacement, I have saved some samples of data for you under [data_sample](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw10-thibodeau-mylinh/data_sample). I have kept the functions below to indicate the code that was used to generate the xml documents and associated datasets. 
```{r}
# get_lots_of_info_caffey <- get_lots_of_info(114000)
# saveXML(get_lots_of_info_caffey, "data_sample/get_lots_of_info_caffey_example2.xml")
read_xml("data_sample/get_lots_of_info_caffey_example2.xml") %>% 
	xmlParse() 
```

A more general httr function for OMIM query.
```{r}
get_omim_search_httr <- 
	function(
	search_term = "coloboma",
	text = FALSE, #Includes the text field sections with the entry.
  	existflags = FALSE, #Include the 'exists' flags with the entry (clinical synopsis, allelic variant, gene map & phenotype map).
  	allelicVariantList = FALSE, #Includes the allelic variant list with the entry.
  	clinicalSynopsis = FALSE, #Include the clinical synopsis with the entry.
  	seeAlso = FALSE, #Includes the 'see also' field with the entry.
  	referenceList = FALSE, #Include the reference list with the entry.
  	geneMap = FALSE, #Include the gene map/phenotype map data with the entry.
  	externalLinks = FALSE, #Include the external links with the entry.
  	contributors = FALSE, #Includes the 'contributors' field with the entry.
  	creationDate = FALSE, #Includes the 'creation date' field with the entry.
  	editHistory = FALSE, #Includes the 'edit history' field with the entry.
  	dates = FALSE, #Include the dates data with the entry.
  	all = FALSE #Include the above data with the entry.
	)
	
	{
	# Obtain the arguments the functions calls to build an URL and get XML
	list_arguments <- as.list(match.call())
  	my_search  <- paste(search_term, '&filter=&fields=&retrieve=&start=0&limit=10&sort=&operator=format=xml', sep='')
  	my_link  <- 'https://api.omim.org/api/entry/search?search='
  	my_query <- glue(my_link, my_search)
  	#loop through all the arguments
  	for (elem in names(list_arguments)){
  		#skip the search_term and blank argument
  		if(!elem %in% '' && !elem %in% 'search_term'){
  			 my_include <- paste('&', 'include=', elem, sep='')
  			 my_query <- paste(my_query, my_include, sep='')
  			 }
  	}
  	my_query_string <- as.character(glue(my_query,"&" ,my_key))
  	print(my_query_string)
  	my_search_query_result <- GET(my_query_string)
	my_search_query_content <- content(my_search_query_result)
	my_search_query_content_parse <- xmlParse(my_search_query_content, useInternalNodes = TRUE)
	return(my_search_query_content_parse)
}
```

So if we take a look at the search for coloboma (default search term).
```{r}
# get_omim_search_httr_coloboma <- get_omim_search_httr() 
# saveXML(get_omim_search_httr_coloboma, "data_sample/get_omim_search_httr_coloboma_example1.xml")
read_xml("data_sample/get_omim_search_httr_coloboma_example1.xml") %>% 
	xmlParse() 
```

Let's just make sure we are able to specify additional terms in the query function `get_omim_search_httr()`, say for example, geneMap for coloboma.
```{r}
# get_omim_search_httr_coloboma_genemap <- get_omim_search_httr(geneMap = TRUE) 
# saveXML(get_omim_search_httr_coloboma_genemap, "data_sample/get_omim_search_httr_coloboma_genemap_example2.xml")
read_xml("data_sample/get_omim_search_httr_coloboma_genemap_example2.xml") %>% 
	xmlParse(useInternalNodes = TRUE) 
```

# FINAL NOTE

The following requirements from hw10 were unclear, so I interpreted them as follow:

* Traverse pages: obtain API data from different API webpages, which has been completed in this homework.
* Send an authorization token: since this assignment is focused on how to get web data with an API and not on writing an API (which is beyond what is covered in this course), I interpret that this requirement refers to "obtaining an authorization token and using it to query an API", which has also been completed in this homework. Note that additional effort has been made to respect the restrictions of the OMIM API and that only a sample of data is provided to the reviewers in line with these restrictions.


***

## THIS WORK HAS BEEN COMPLETED FOR LEARNING PURPOSES ONLY

## SOURCES OF DATA - REFERENCES

* Online Mendelian Inheritance in Man, OMIM®. McKusick-Nathans Institute of Genetic Medicine, Johns Hopkins University (Baltimore, MD), November 29 2017. World Wide Web URL: https://omim.org/
* Code adapted from Dave Tang tutorial Getting started with the OMIM api [here](https://davetang.org/muse/2015/03/17/getting-started-with-the-omim-api/)
* DECIPHER: Database of Chromosomal Imbalance and Phenotype in Humans using Ensembl Resources. Firth, H.V. et al (2009). Am.J.Hum.Genet 84, 524-533 (DOI: dx.doi.org/10/1016/j.ajhg.2009.03.010). 
* This homework makes use of genomic data generated by the DECIPHER community. A full list of centres who contributed to the generation of the data is available from http://decipher.sanger.ac.uk and via email from decipher@sanger.ac.uk. Funding for the project was provided by the Wellcome Trust.
* Gray KA, Yates B, Seal RL, Wright MW, Bruford EA. genenames.org: the HGNC resources in 2015. Nucleic Acids Res. 2015 Jan;43(Database issue):D1079-85. doi: 10.1093/nar/gku1071. PMID:25361968. 
* HGNC Database, HUGO Gene Nomenclature Committee (HGNC), EMBL Outstation - Hinxton, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridgeshire, CB10 1SD, UK www.genenames.org (complete HGNC dataset downloaded [here](https://www.genenames.org/cgi-bin/statistics) on November 17/2017).
* Orphanet: an online database of rare diseases and orphan drugs. Copyright, INSERM 1997. Available at http://www.orpha.net Accessed (accessed November 17, 2017).
* Orphadata: Free access data from Orphanet. © INSERM 1997. Available on http://www.orphadata.org. Data version (XML data version en_product4_HPO.xml).


## GENERAL REFERENCES

* stat545 webdata03 lecture 2015 [here](http://stat545.com/webdata03_activity.html)
* collapse a vector into a string [here](http://r.789695.n4.nabble.com/Concatenating-one-character-vector-into-one-string-td835795.html)
* capture output use [here](https://stackoverflow.com/questions/27594541/export-a-list-into-a-csv-or-txt-file-in-r)

