# README

stat547-hw10-thibodeau-mylinh

## HOMEWORK 10 

In this homework, I will be working with the Online Mendelian Inheritance of Man - [OMIM](https://omim.org/) data as well as previously used merged data from [DECIPHER](http://decipher.sanger.ac.uk), [Orphanet](http://www.orpha.net) and [HGNC](https://www.genenames.org/cgi-bin/statistics) (see references and [homework 8](https://github.com/mylinhthibodeau/STAT545-HW-thibodeau-mylinh/tree/master/stat547-hw8-thibodeau-mylinh) for more information)

Extensive information on the OMIM API is available [here](https://omim.org/help/api), but I will provide a few examples of query and functions to query the API.

**IMPORTANT:** This work is adapted from Dave Tang's tutorial on Getting started with the OMIM api [here](https://davetang.org/muse/2015/03/17/getting-started-with-the-omim-api/).

**IMPORTANT:** An alternative method has been used to retrieve the OMIM API data: I mainly used adapted functions instead of `GET()` from the library httr.

**IMPORTANT:** I spent quite a bit of time trying to make some output files that you guys could see while respecting the protected data restriction of the OMIM API. Please install the packages XML and xml2 in order to be able to take a better look at the OMIM API xml output examples in the [data_sample folder]("stat547-hw10-thibodeau-mylinh/data_sample"), and otherwise, I have put some data samples in the homework, but the aesthetic of it is not great. 

# Homework 10 requirements

Here is the breakdown of requirements and corresponding sections:

* Get data from API - Section A (part 1, 2, 3, 4, 5) and Section D (part 1) 
* Clean and tidy into dataframes - Section B (part 2) and Section C and Section D (part 2)
* Save output to files
	- protected data has been saved under the data folder - not public (e.g. mim2gene_and_inheritance_DDG2P_hgnc_orphadata.tsv)
	- sample data for reviewers saved under [data_sample folder]("stat547-hw10-thibodeau-mylinh/data_sample")
* Use httr and `GET()` - Section F contains some examples of API query using httr. A new function using httr and glue is defined.

The following requirements from hw10 were unclear, so I interpreted them as follow:

* Traverse pages: obtain API data from different API webpages, which has been completed in this homework.
* Send an authorization token: since this assignment is focused on how to get web data with an API and not on writing an API (which is beyond what is covered in this course), I interpret that this requirement refers to "obtaining an authorization token and using it to query an API", which has also been completed in this homework. Note that additional effort has been made to respect the restrictions of the OMIM API and that only a sample of data is provided to the reviewers in line with these restrictions.

*Please note that I have not used purrr to 

### Challenges encountered

#### Example 1


#### Example 2


#### Example 3


### Open questions

### New skills acquired

1. Store API key safely and privately in an .Rprofile file and add the file name to .gitignore file (so that the file will not be pushed to github). This is a way to keep an API key private. 
2. Use an API to make web data query: format of query and particularities relating to the OMIM API format.
3. Write functions to make API data query: main code was adapted from Dave Tang tutorial Getting started with the OMIM api [here](https://davetang.org/muse/2015/03/17/getting-started-with-the-omim-api/) for main body of the homework, but a I wrote a new function that builds on previous work and uses httr and `GET()` in the final section of the homework.
4. 

***

# References

## THIS WORK HAS BEEN COMPLETED FOR LEARNING PURPOSES ONLY

## SOURCES OF DATA - REFERENCES

* Online Mendelian Inheritance in Man, OMIM®. McKusick-Nathans Institute of Genetic Medicine, Johns Hopkins University (Baltimore, MD), November 29 2017. World Wide Web URL: https://omim.org/
* Code adapted from Dave Tang tutorial Getting started with the OMIM api [here](https://davetang.org/muse/2015/03/17/getting-started-with-the-omim-api/)
* DECIPHER: Database of Chromosomal Imbalance and Phenotype in Humans using Ensembl Resources. Firth, H.V. et al (2009). Am.J.Hum.Genet 84, 524-533 (DOI: dx.doi.org/10/1016/j.ajhg.2009.03.010). 
* This homework makes use of genomic data generated by the DECIPHER community. A full list of centres who contributed to the generation of the data is available from http://decipher.sanger.ac.uk and via email from decipher@sanger.ac.uk. Funding for the project was provided by the Wellcome Trust.
* Gray KA, Yates B, Seal RL, Wright MW, Bruford EA. genenames.org: the HGNC resources in 2015. Nucleic Acids Res. 2015 Jan;43(Database issue):D1079-85. doi: 10.1093/nar/gku1071. PMID:25361968. 
* HGNC Database, HUGO Gene Nomenclature Committee (HGNC), EMBL Outstation - Hinxton, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridgeshire, CB10 1SD, UK www.genenames.org (complete HGNC dataset downloaded [here](https://www.genenames.org/cgi-bin/statistics) on November 17/2017).
* Orphanet: an online database of rare diseases and orphan drugs. Copyright, INSERM 1997. Available at http://www.orpha.net Accessed (accessed November 17, 2017).
* Orphadata: Free access data from Orphanet. © INSERM 1997. Available on http://www.orphadata.org. Data version (XML data version en_product4_HPO.xml).


## GENERAL REFERENCES

* stat545 webdata03 lecture 2015 [here](http://stat545.com/webdata03_activity.html)
* collapse a vector into a string [here](http://r.789695.n4.nabble.com/Concatenating-one-character-vector-into-one-string-td835795.html)
* capture output use [here](https://stackoverflow.com/questions/27594541/export-a-list-into-a-csv-or-txt-file-in-r)
